File: php/src/Amen/Resources/Deals.php
<?php
declare(strict_types=1);
namespace Amen\Resources;
use Amen\{AmenClient, AmenLifecycleError};

/** POST /deals/{n}/action/* — every method returns the updated deal (or a checkout for online payment). */
final class DealActions
{
    /** Which statuses each action may be called from (docs/02-deal-lifecycle.md). */
    public const ALLOWED_FROM = [
        'submit' => ['draft'], 'approve' => ['requested'],
        'make-payment-wallet' => ['payment_pending'], 'make-payment-online' => ['payment_pending'],
        'execution-start' => ['paid'], 'execution-complete' => ['executing'], 'complete' => ['executed'],
        'transfer-seller-amount' => ['completed'], 'dispute' => ['completed'],
        'dispute-approve' => ['disputed'], 'dispute-decline' => ['disputed'],
        'cancel' => ['draft', 'requested', 'payment_pending', 'paid', 'executing'],
    ];
    public function __construct(private AmenClient $c, private Deals $deals) {}

    private function act(string $n, string $action, array $opts = [], bool $check = true): array
    {
        if ($check) {
            $status = $this->deals->get($n)['status'] ?? null;
            if (!in_array($status, self::ALLOWED_FROM[$action], true))
                throw new AmenLifecycleError("action '$action' is not allowed from status '$status' (allowed: " . implode(', ', self::ALLOWED_FROM[$action]) . ')');
        }
        return $this->c->request('POST', "/deals/$n/action/$action", $opts) ?? [];
    }
    public function submit(string $n, bool $check = true): array { return $this->act($n, 'submit', [], $check); }
    public function approve(string $n, ?string $price = null, bool $check = true): array { return $this->act($n, 'approve', ['json' => $price ? ['price' => $price] : new \stdClass()], $check); }
    public function payWithWallet(string $n, bool $check = true): array { return $this->act($n, 'make-payment-wallet', [], $check); }
    public function payOnline(string $n, string $paymentMethod = 'mada', bool $check = true): array { return $this->act($n, 'make-payment-online', ['json' => ['payment_method' => $paymentMethod]], $check); }
    public function executionStart(string $n, bool $check = true): array { return $this->act($n, 'execution-start', [], $check); }
    public function executionComplete(string $n, bool $check = true): array { return $this->act($n, 'execution-complete', [], $check); }
    public function complete(string $n, bool $check = true): array { return $this->act($n, 'complete', [], $check); }
    public function transferSellerAmount(string $n, bool $check = true): array { return $this->act($n, 'transfer-seller-amount', [], $check); }
    public function cancel(string $n, string $dealParty, int $reason, string $comment, bool $check = true): array
    { return $this->act($n, 'cancel', ['json' => ['deal_party' => $dealParty, 'reason' => $reason, 'comment' => $comment]], $check); }
    /** @param string[] $attachmentPaths */
    public function dispute(string $n, int $reason, string $comment, array $attachmentPaths = [], bool $check = true): array
    { $form = ['reason' => $reason, 'comment' => $comment]; foreach ($attachmentPaths as $i => $p) $form['attachment_' . ($i + 1)] = new \CURLFile($p); return $this->act($n, 'dispute', ['form' => $form], $check); }
    public function disputeApprove(string $n, int $reason, string $comment, bool $check = true): array { return $this->act($n, 'dispute-approve', ['form' => ['reason' => $reason, 'comment' => $comment]], $check); }
    public function disputeDecline(string $n, int $reason, string $comment, bool $check = true): array { return $this->act($n, 'dispute-decline', ['form' => ['reason' => $reason, 'comment' => $comment]], $check); }
}

final class Deals
{
    public readonly DealActions $actions;
    public function __construct(private AmenClient $c) { $this->actions = new DealActions($c, $this); }
    public function create(array $body): array { return $this->c->request('POST', '/deals/', ['json' => $body]); }
    public function get(string $n): array { return $this->c->request('GET', "/deals/$n"); }
    public function update(string $n, array $body): array { return $this->c->request('PUT', "/deals/$n", ['json' => $body]); }
    public function delete(string $n): void { $this->c->request('DELETE', "/deals/$n"); }
    public function list(array $params = []): array { return Page::from($this->c->request('GET', '/deals/', ['params' => $params]), 'deals'); }
    public function iterAll(array $filters = []): \Generator
    { for ($page = 0; ; $page++) { $p = $this->list($filters + ['page' => $page]); yield from $p['items']; if ($page + 1 >= $p['pages'] || !$p['items']) return; } }
    public function setParties(string $n, array $buyers, array $sellers): array { return $this->c->request('POST', "/deals/$n/parties/", ['json' => ['buyers' => $buyers, 'sellers' => $sellers]]); }
    public function setDeliveryAddress(string $n, array $address): array { return $this->c->request('POST', "/deals/$n/delivery-address", ['json' => $address]); }
    public function setBillingAddress(string $n, array $address): array { return $this->c->request('POST', "/deals/$n/billing-address", ['json' => $address]); }
    public function allowedPaymentMethods(string $n): array { return $this->c->request('GET', "/deals/$n/allowed-payment-methods/")['payment_methods'] ?? []; }
}

File: python/src/amen/resources/deals.py
from __future__ import annotations
from typing import Iterator
from ..models import Deal, Page, Checkout
from ..errors import AmenLifecycleError

# Which deal statuses each action may be called from (docs/02-deal-lifecycle.md)
ALLOWED_FROM = {
    "submit": {"draft"}, "approve": {"requested"},
    "make-payment-wallet": {"payment_pending"}, "make-payment-online": {"payment_pending"},
    "execution-start": {"paid"}, "execution-complete": {"executing"}, "complete": {"executed"},
    "transfer-seller-amount": {"completed"}, "dispute": {"completed"},
    "dispute-approve": {"disputed"}, "dispute-decline": {"disputed"},
    "cancel": {"draft", "submitted", "requested", "payment_pending", "paid", "executing"},
}


class DealActions:
    """POST /deals/{n}/action/* — every method returns the updated Deal (or a Checkout for online payment)."""

    def __init__(self, client, deals): self._c, self._deals = client, deals

    def _act(self, deal_number: str, action: str, *, json=None, data=None, files=None, check: bool = True) -> dict:
        if check:
            status = self._deals.get(deal_number).status
            if status not in ALLOWED_FROM[action]:
                raise AmenLifecycleError(
                    f"action '{action}' is not allowed from status '{status}' (allowed: {sorted(ALLOWED_FROM[action])})")
        return self._c.request("POST", f"/deals/{deal_number}/action/{action}", json=json, data=data, files=files)

    def submit(self, n, check=True) -> Deal: return Deal.from_api(self._act(n, "submit", check=check))
    def approve(self, n, price: str | None = None, check=True) -> Deal:
        return Deal.from_api(self._act(n, "approve", json={"price": price} if price else {}, check=check))
    def pay_with_wallet(self, n, check=True) -> Deal: return Deal.from_api(self._act(n, "make-payment-wallet", check=check))
    def pay_online(self, n, payment_method: str = "mada", check=True) -> Checkout:
        return Checkout.from_api(self._act(n, "make-payment-online", json={"payment_method": payment_method}, check=check))
    def execution_start(self, n, check=True) -> Deal: return Deal.from_api(self._act(n, "execution-start", check=check))
    def execution_complete(self, n, check=True) -> Deal: return Deal.from_api(self._act(n, "execution-complete", check=check))
    def complete(self, n, check=True) -> Deal: return Deal.from_api(self._act(n, "complete", check=check))
    def transfer_seller_amount(self, n, check=True) -> Deal: return Deal.from_api(self._act(n, "transfer-seller-amount", check=check))
    def cancel(self, n, *, deal_party: str, reason: int, comment: str, check=True) -> Deal:
        return Deal.from_api(self._act(n, "cancel", json={"deal_party": deal_party, "reason": reason, "comment": comment}, check=check))
    def dispute(self, n, *, reason: int, comment: str, attachments: list[tuple] | None = None, check=True) -> Deal:
        files = {f"attachment_{i+1}": a for i, a in enumerate(attachments or [])} or {"_": (None, "")}
        return Deal.from_api(self._act(n, "dispute", data={"reason": reason, "comment": comment}, files=files, check=check))
    def dispute_approve(self, n, *, reason: int, comment: str, check=True) -> Deal:
        return Deal.from_api(self._act(n, "dispute-approve", data={"reason": reason, "comment": comment}, files={"_": (None, "")}, check=check))
    def dispute_decline(self, n, *, reason: int, comment: str, check=True) -> Deal:
        return Deal.from_api(self._act(n, "dispute-decline", data={"reason": reason, "comment": comment}, files={"_": (None, "")}, check=check))


class Deals:
    def __init__(self, client):
        self._c = client
        self.actions = DealActions(client, self)

    def create(self, *, offer_type: str, offer_title: str, offer_price: str | None = None, offer_category: int | None = None,
               offer_description: str | None = None, offer_delivery_fee: str | None = None,
               deal_subject_details: str | None = None) -> Deal:
        body = {k: v for k, v in locals().items() if k != "self" and v is not None}
        return Deal.from_api(self._c.request("POST", "/deals/", json=body))

    def get(self, deal_number: str) -> Deal:
        return Deal.from_api(self._c.request("GET", f"/deals/{deal_number}"))

    def update(self, deal_number: str, **fields) -> Deal:
        return Deal.from_api(self._c.request("PUT", f"/deals/{deal_number}", json=fields))

    def delete(self, deal_number: str) -> None:
        self._c.request("DELETE", f"/deals/{deal_number}")

    def list(self, *, page: int | None = None, per_page: int | None = None, status: str | None = None) -> Page:
        params = {k: v for k, v in {"page": page, "per_page": per_page, "status": status}.items() if v is not None}
        return Page.from_api(self._c.request("GET", "/deals/", params=params), "deals")

    def iter_all(self, **filters) -> Iterator[Deal]:
        page = 0
        while True:
            p = self.list(page=page, **filters)
            yield from (Deal.from_api(i) for i in p.items)
            page += 1
            if page >= p.pages or not p.items:
                return

    def set_parties(self, deal_number: str, *, buyers: list[str], sellers: list[str]) -> Deal:
        return Deal.from_api(self._c.request("POST", f"/deals/{deal_number}/parties/", json={"buyers": buyers, "sellers": sellers}))

    def set_delivery_address(self, deal_number: str, *, city: int, street: str, building_number: str, zip_code: str,
                             district: str | None = None, unit_number: str | None = None) -> Deal:
        body = {k: v for k, v in locals().items() if k not in ("self", "deal_number") and v is not None}
        return Deal.from_api(self._c.request("POST", f"/deals/{deal_number}/delivery-address", json=body))

    def set_billing_address(self, deal_number: str, **address) -> Deal:
        return Deal.from_api(self._c.request("POST", f"/deals/{deal_number}/billing-address", json=address))

    def allowed_payment_methods(self, deal_number: str) -> list[str]:
        return self._c.request("GET", f"/deals/{deal_number}/allowed-payment-methods/").get("payment_methods", [])
